server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAO.java

```diff
@@ -177,4 +188,16 @@ public PayloadCodec getPayloadCodec() {
     public Optional<String> getBucketPrefix() {
         return bucketPrefix;
     }
+
+    private BucketName resolveToBackendBucketName(BucketName bucketName) {
+        Preconditions.checkNotNull(bucketName);
+
+        if (bucketName.equals(defaultBucketName)) {
```

Comment:
That, is not what we said, sadly...

We said:
 - if namespace is configured, use it without prefix (for retro-compatibility)
 - if namespace is empty, then we use BucketName.DEFAULT, and we should apply the prefix (for multi-tenant)
 - if namespace is configured to BucketName.DEFAULT, we should not apply the prefix (retro-compatibility edge case)

Quick note: 
 - You need to have the configured namespace (optional) and not the resolved value. Please replace `BucketName defaultBucketName` field by an `Optional<BucketName> namespace` value.
 - You may miss unit tests here.

Reply:
- if namespace is configured, use it without prefix (for retro-compatibility)
 - if namespace is configured to BucketName.DEFAULT, we should not apply the prefix (retro-compatibility edge case)

=> if namespace is configured, keep theDefaultBucketName as it is, I did it

 - if namespace is empty, then we use BucketName.DEFAULT, and we should apply the prefix (for multi-tenant)

=> I did it.

Please checkout the ObjectStorageBlobsDAOBuilder, that is where I wrote this:
```java
        public ObjectStorageBlobsDAO build() {
            ...
            BucketName defaultBucketName = this.defaultBucketName
                .orElseGet(() -> defaultWithPrefix(bucketPrefix));

            return new ObjectStorageBlobsDAO(defaultBucketName,
                blobIdFactory,
                blobStore,
                putBlob.orElse(new DefaultPutBlobFunction(blobStore)),
                payloadCodec.orElse(PayloadCodec.DEFAULT_CODEC),
                bucketPrefix);
        }

        private BucketName defaultWithPrefix(Optional<String> bucketPrefix) {
            return bucketPrefix.map(BucketName.DEFAULT::withPrefix)
                .orElse(BucketName.DEFAULT);
        }
```

Reply:
and test I wrote to cover it is in AwsS3ObjectStorageBlobsDAOBuilderTest 91177c26a6db60d3541604bfa1ad2aeb9838b271

Then the resolving logic is simpler in ObjectStorageBlobsDAO

Reply:
Pfff I'm not satified with the tests that do not allow a quick review of the rBucketName resolution algorithm, that as you just said it, span several classes.

We should clearly establish the responsability: please extract a `BackendBucketNameResolver` that we could quicky test without requiring a all docker and a builder hell. Everything related with translation should be located there and we should hunt edge cases down.

Reply:
Ps: I don't understand that `BucketName defaultWithPrefix(Optional<String> bucketPrefix)` stuff. Does that mean that the prefix is applied in the builder for defaultBucketName (and thus used by the caller) while for custom bucket is is applied in the BlobStore?

That's a strong design smell: we need such decisions to be centralized.

Also I bet typing is not good enough here... You are mixing concerns together.

Please introduce a new `ObjectStorageBucketName` class.

 - The caller uses an applicative level `BucketName` request
 - The blobStore translates it in an `ObjectStorageBucketName`
 - Lower lever storage API's within ObjectStrage BlobStore implem would require `ObjectStorageBucketName`

Not only does this prevent from 'forgetting' to do the translation, but also it will force you to apply a DRY, SRP logic here.